### PR TITLE
[ELF] Assign the correct SHNDX to symbols killed by ICF.

### PR DESCRIPTION
**Pull request by ishitatsuyuki**
We have been setting the symbol value and size correctly, but the SHNDX was
erroneously set to UND, potentially causing undefined symbol errors. Fix this by
forwarding the appropriate value from the `leader`.

Closes: #779

### CODE REVIEW

**Comment by ishitatsuyuki** [comment]
Draft until the fix is confirmed by #779's reporter.

File: elf/mold.h [hunk 1]
```diff
@@ -329,6 +329,13 @@ class InputSection {
   std::atomic_bool is_visited = false;
 
   // For ICF
+  //
+  // `leader` is the section that this section has been merged with.
+  // Three kind of values are possible:
+  // - `leader == nullptr`: This section was not eligible for ICF.
+  // - `leader == this`: This section was retained.
+  // - `leader != this`: This section was merged with another idential section.
+  //                     Implies `killed_by_icf`.
```
Comment (rui314) [inline elf/mold.h:338]: Does this mean we can get rid of `killed_by_icf` flag and use this pointer instead?

**Comment by rui314** [comment]
Great, thanks!

File: elf/mold.h (hunk 1 above)
Comment (ishitatsuyuki) [reply elf/mold.h:338]: It’s assigned [here](https://github.com/rui314/mold/blob/b3ed06aa3956ed309ad2fe9ab757f6ee87f4f467/elf/icf.cc#L598), and it’s assigned from a single place, so we should be able to replace this with a helper function denoting the same condition.

Comment (rui314) [reply elf/mold.h:338]: Do you mind if I ask you to do this?

Comment (ishitatsuyuki) [reply elf/mold.h:338]: No problem, will do.

**Pull request merged**